### ISSUE TITLE
[es 1.4.4] Scroll query returns 0 results before fully done 

### ISSUE THREAD

**Issue by hjz** [elastic/elasticsearch#10417]
We're on ES 1.4.4, running a 16 node cluster +  3 dedicated client nodes (c4.xlarge) which runs all our queries.

We're seeing intermittent failures on scroll queries on larger data sets of ~1 million documents. This has happened while we were on 1.3.2 as well.

Today, it failed after fetching 5000 documents (5 requests, each fetch is 1000 documents), then started returning 0. Retrying the scroll query again worked until completion. Each fetch took 1-2 seconds before failure and then started returning almost immediately with 0 results.

This index in particular has 5 shards and 26gb of data total.

Scroll size per shard: 200
Scroll time: 4 minutes

Let me know if there's other information I can provide to help diagnose

**Comment by jpountz** [comment]
When the query returns 0 results prematurely, do you have shard failures in the response?

**Comment by hjz** [comment]
We've seen shard failures before where 1 or 2 shards would fail and instead of 1000 results, it'll return 900 or 800. 

Didn't log the shard failures this time, suppose all 5 shards failed, any idea why that would happen? I'll add logging for that and circle back.

**Comment by clintongormley** [comment]
@hjz can you provide the actual code you're using for scroll requests, plus any exceptions?

**Comment by hjz** [comment]
Here's the query to start the scroll

```scala
  def startScrollQueryGetTotalHitsAndScrollId(index: String, filter: BaseFilterBuilder): (Long, String) = {
    val r = ES.client.prepareSearch(index)
      .setTypes(User.eventType)
      .setSearchType(SearchType.SCAN)
      .setFrom(0)
      .setSize(config.scrollSizePerShard)
      .setScroll(scrollTime)

     r.setQuery(QueryBuilders.filteredQuery(QueryBuilders.matchAllQuery(), filter))

    val resp = r.execute().actionGet(config.scrollTimeoutSecs, TimeUnit.SECONDS)

    (resp.getHits.getTotalHits, resp.getScrollId)
  }
```

Then subsequent looks up are done via:

```scala
      val scrollResp = ES.client
        .prepareSearchScroll(scrollId)
        .setScroll(scrollTime)
        .execute()
        .actionGet(config.scrollTimeoutSecs, TimeUnit.SECONDS)

      if (scrollResp.getFailedShards > 0) {
        val shards = scrollResp.getShardFailures
        shards.foreach { s =>
          val msg = s"Shard failed scroll query: ${s.index()}, shardId: ${s.shardId()} reason: ${s.reason()}}"
          hipChatService.errorMessage(msg, createAlert = false)
        }
      }

      (scrollResp.getHits.getHits.map { sh =>
        ApiUser.fromSource(sh.getId, sh.source() )
      }.toSeq,  scrollResp.getScrollId)
```

I'll follow up with exceptions

**Comment by clintongormley** [comment]
@hjz I don't see where you are getting the scroll_id from the previous scroll request and using it for the next scroll request.

Normally there would be some `while` loop which does a scroll request, checks if there are any hits, then issues a new scroll request, using the `scroll_id` returned by the previous scroll request.

**Comment by hjz** [comment]
Yes that what we do. I omittied some code there. Essentially

```scala
def getNextScroll() = {
      val scrollResp = ES.client
        .prepareSearchScroll(scrollId)
        .setScroll(scrollTime)
        .execute()
        .actionGet(config.scrollTimeoutSecs, TimeUnit.SECONDS)

      if (scrollResp.getFailedShards > 0) {
        val shards = scrollResp.getShardFailures
        shards.foreach { s =>
          val msg = s"Shard failed scroll query: ${s.index()}, shardId: ${s.shardId()} reason: ${s.reason()}}"
          hipChatService.errorMessage(msg, createAlert = false)
        }
      }

      (scrollResp.getHits.getHits.map { sh =>
        ApiUser.fromSource(sh.getId, sh.source() )
      }.toSeq,  scrollResp.getScrollId)
}
```

```scala
val (totalHits, startScrollId) = startScrollQueryGetTotalHitsAndScrollId
var currentScrollId = startScrollId
    var batch = 1
    var remaining = totalHits
    var retries = 0
    var batchHits = 0L

do {
      val startTime = DateTime.now

     getNextScroll(currentScrollId) match {
        case Success((users, nextScrollId)) =>
          currentScrollId = nextScrollId
          batch += 1

        case Failure(e) =>
          batchHits = 0
          hipChatService.sendMessage(s"$logHdr scroll query error. ${e.getMessage}. retry $retries", color = HipChatNotificationColor.Red)
      }

      // If hits was 0 but there are users remaining, something fubared, so sleep for BatchRetrySeconds and retry
      if (batchHits == 0 && remaining > 0) {
        retries += 1
        Logger.error(s"$logHdr Elasticsearch scroll request failed, retries: $retries. Sleeping for ${config.scrollRetrySleepDuration.seconds} seconds")
        Thread.sleep(config.scrollRetrySleepDuration.millis)
      }

      remaining -= batchHits

    } while (remaining > 0 && retries < config.scrollMaxRetries)
```

**Comment by pickypg** [comment]
@hjz Have you gotten any logs for this issue yet?

**Comment by hjz** [comment]
Looks like the cause is:

2015-04-20 17:32:14,633 [ERROR] application - Shard failed scroll query: null, shardId: -1 reason: RemoteTransportException[[xxx][inet[/xxxx]][indices:data/read/search[phase/scan/scroll]]]; nested: EsRejectedExecutionException[rejected execution (queue capacity 1000) on org.elasticsearch.transport.netty.MessageChannelHandler$RequestHandler@4546354d]; }

Good call on logging the shard failures. Is there a way to always let scroll queries succeed?

**Comment by pickypg** [comment]
@hjz For your value of `index`, are you using `"_all"` (or equivalent) by chance? How many shards are actually associated with that request?

**Comment by hjz** [comment]
We have _all turned off. This is hitting 15 shards that has 2 replicas.

**Comment by l15k4** [comment]
I seem to have this issue too... I'm scanning a little more than 100 indices with `keepAlive=15s` : 
```
            "number_of_replicas": "1",
            "number_of_shards": "5",
```
on : 
```
   "number_of_nodes": 4,
   "number_of_data_nodes": 4,
   "active_primary_shards": 530,
   "active_shards": 1060,
```
I haven't noticed it would happen with sorted `QueryThenFetch` scroll because I couldn't scroll as much as I can `Scan`. But when scanning it returns 0 results after ~ 100M documents from 350M. I don't check `SearchResponse#getShardFailures` though. What should I do when there is a failure? Repeat the request with the same `scroll_id` ?

**Comment by l15k4** [comment]
I figured out why this happened to me, the reason is that scanning slows down so much that it gets out of `10s` keepAlive period which leads to `SearchResponse#shardFailures`. These logs document it :

**Comment by vvcephei** [comment]
I have had some trouble with this as well. Sadly, it's been long enough that I won't be able to remember the details for debugging. (I just happened to see this issue while looking for something else.)

What I can say is that I've been able to (as far as I can tell) eliminate the possibility of missing data during a scroll by counting the number of documents I've seen and checking at the end of the scroll that it _exactly_ matches the total_hits reported by the initial query.

This seems like a good sanity check for anyone doing a scan&scroll, as there's no reason for the hits to differ from the total_hits.

**Comment by clintongormley** [comment]
The original issue looks like the cluster was overloaded. I don't think there is anything to do here.  Closing

**Comment by vvcephei** [comment]
Hey Clinton,

Yeah, our clusters are generally pretty heavily loaded, but I don't know
how to quantify it being overloaded.

This error actually happens to us all the time, and we simply detect it and
start the scroll again.

Thanks,
John

On Friday, January 15, 2016, Clinton Gormley <notifications@github.com>
wrote:
> Closed #10417.
>
> —
> Reply to this email directly or view it on GitHub.<
https://ci5.googleusercontent.com/proxy/D3fUvDCaH68UfTcyrC_41gEBYmFovoX9yBs4r5lq1Laf4bxsDbQSWk6_se-Oi4oNSPW8HJJdyOFXj7r0ZdCsnJNsF2dhmKquDRV0JrTGA0sxdrSwRx6x1s1ATQgnOXp5fiJiFvulaNuD4TduirJrHx3lOHzZMA=s0-d-e1-ft#https://github.com/notifications/beacon/AAy1E4J4EKL3Uyv3_42vkojmEbfE-RTrks5paT2FgaJpZM4D537e.gif
>

**Comment by clintongormley** [comment]
@vvcephei see these in the logs from the https://github.com/elastic/elasticsearch/issues/10417#issuecomment-94538026:

>  [indices:data/read/search[phase/scan/scroll]]]; nested: EsRejectedExecutionException[rejected execution (queue capacity 1000)

You can keep an eye on your search thread pool queue size.  If you're regularly filling it up then either you need (a) more efficient queries or (b) more hardware.
